pkg/reconciler/route/route.go

```diff
@@ -276,9 +277,45 @@ func (c *Reconciler) tls(ctx context.Context, host string, r *v1.Route, traffic
 	sort.Slice(acmeChallenges, func(i, j int) bool {
 		return acmeChallenges[i].URL.String() < acmeChallenges[j].URL.String()
 	})
+
+	orphanCerts, err := c.getOrphanRouteCerts(r, domainToTagMap)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, cert := range orphanCerts {
+		c.GetNetworkingClient().NetworkingV1alpha1().Certificates(cert.Namespace).Delete(ctx, cert.Name, metav1.DeleteOptions{})
+	}
+
 	return tls, acmeChallenges, nil
 }
 
+// Returns a slice of certificates that used to belong route's old tags and are currently not in use.
+func (c *Reconciler) getOrphanRouteCerts(r *v1.Route, domainToTagMap map[string]string) ([]*netv1alpha1.Certificate, error) {
+	var unusedCerts []*netv1alpha1.Certificate
+	labelSelector := kubelabels.SelectorFromSet(kubelabels.Set{
+		serving.RouteLabelKey: r.Name,
+	})
+
+	certs, err := c.certificateLister.Certificates(r.Namespace).List(labelSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	var shouldKeepCert bool
+	for _, cert := range certs {
+		for _, dn := range cert.Spec.DNSNames {
+			_, shouldKeepCert = domainToTagMap[dn]
+		}
+
+		if !shouldKeepCert {
+			unusedCerts = append(unusedCerts, cert)
+		}
+	}
```

Comment:
I think there might be a little bug here: if the first cert matches, but the second doesn't `shouldKeepCert` will be false even though it should be true (it'll be overwritten). We probably want something like the following (untested!), and it'd be good to add a test case for this situation.

```suggestion
	for _, cert := range certs {
		var shouldKeepCert bool
		for _, dn := range cert.Spec.DNSNames {
			if _, used = domainToTagMap[dn]; used {
			  shouldKeepCert = true
			}
		}

		if !shouldKeepCert {
			unusedCerts = append(unusedCerts, cert)
		}
	}
```